### PR TITLE
Missing Data Fix, Partial Keys Fix

### LINKED ISSUE #204: [BUG] Misleading Mixed Data Structure Outputs
# Expected Behavior
When joining mixed data structures with `read_redcap(..., allow_mixed_structure = TRUE)`, we expect tables to be made with meaningful, unique primary keys.

# Current Behavior
During development of #199, we found that joining tables that both "repeat together" (i.e. as events, RT) and "repeat separate" (i.e. as instruments, RS) don't provide the appropriate primary keys to allow for distinction of RS rows during joins.

## How to Reproduce the Bug:
Using the REDCap for a single record as set up below:
image

Currently `read_redcap()` gives us the following output:

```r
> sprtbl$redcap_data
[[1]]
# A tibble: 4 × 5
  record_id redcap_event       redcap_form_instance mixed_structure_1 form_status_complete
      <dbl> <chr>                             <dbl> <chr>             <fct>               
1         1 repeating_together                    1 RT1               Complete            
2         1 repeating_together                    2 RT2               Complete            
3         1 repeating_separate                    1 RS1               Complete            
4         1 repeating_separate                    2 RS2               Complete            

[[2]]
# A tibble: 3 × 5
  record_id redcap_event       redcap_form_instance mixed_structure_2 form_status_complete
      <dbl> <chr>                             <dbl> <chr>             <fct>               
1         1 repeating_together                    1 RT1               Complete            
2         1 repeating_together                    2 RT2               Complete            
3         1 repeating_separate                    1 RS1               Complete   
```

Using `join_data_tibbles()` with a "full join" we get this:

```r
join_data_tibbles(sprtbl, x = "mixed_structure_1", y = "mixed_structure_2", type = "full")
# A tibble: 4 × 8
  record_id redcap_event       redcap_form_instance mixed_structure_1 redcap_event_instance mixed_structure_2 form_status_complete.x form_status_complete.y
      <dbl> <chr>                             <dbl> <chr>             <lgl>                 <chr>             <fct>                  <fct>                 
1         1 repeating_together                    1 RT1               NA                    RT1               Complete               Complete              
2         1 repeating_together                    2 RT2               NA                    RT2               Complete               Complete              
3         1 repeating_separate                    1 RS1               NA                    RS1               Complete               Complete              
4         1 repeating_separate                    2 RS2               NA                    NA                Complete               NA        
```

The issue here is that in row 3, data for `mixed_structure_1` and `mixed_structure_2` should exist on separate rows because they are RS instances. As `read_redcap()` is currently set up, it is impossible to separate these because the primary keys for both are identical (`record_id`, `redcap_event`, `redcap_form_instance`). This is a by product of how we decided to mix `redcap_form_instance`s meaning between repeat events and instruments

## Solution Proposal

To fix this we will need to do the following:

- Add the repeating event type to the `redcap_events` column of the supertibble. This will need to identify if an event is a RS/RT for us to reference in other functions like `join_data_tibbles()`
- Revise how `redcap_form_instance` is used in these examples for RS rows. This may involve revising how `redcap_form_instance`/`redcap_event_instance` is defined but boils down to needing an additional primary key to help identify separate repeating instances. @ezraporter after making an example I think `redcap_form_instance`/`redcap_event_instance` might still not address this fully since these share the same event information. I think instead we need to identify RS data and then give it an additional column with the form name it came from. This might be able to get shifted back to the join function's responsibilities.

```r
> tibble::tribble(
+   ~"record_id", ~"redcap_event", ~"redcap_form_instance", ~"redcap_event_instance", ~"extra_rs_key", ~"mixed_structure_1", ~"mixed_structure_2", ~"form_status_complete.x", ~"form_status_complete.y",
+   1, "repeat_together", 1, NA, NA, "RT1", "RT1", "Complete", "Complete",
+   1, "repeat_together", 2, NA, NA, "RT2", "RT2", "Complete", "Complete",
+   1, "repeat_separate", 1, NA, "mixed_structure_1", "RS1", NA, "Complete", NA,
+   1, "repeat_separate", 1, NA, "mixed_structure_2", NA, "RS1", NA, "Complete",
+   1, "repeat_separate", 2, NA, "mixed_structure_1", "RS2", NA, "Complete", NA
+   )
# A tibble: 5 × 9
  record_id redcap_event    redcap_form_instance redcap_event_instance extra_rs_key      mixed_structure_1 mixed_structure_2 form_status_complete.x form_status_complete.y
      <dbl> <chr>                          <dbl> <lgl>                 <chr>             <chr>             <chr>             <chr>                  <chr>                 
1         1 repeat_together                    1 NA                    NA                RT1               RT1               Complete               Complete              
2         1 repeat_together                    2 NA                    NA                RT2               RT2               Complete               Complete              
3         1 repeat_separate                    1 NA                    mixed_structure_1 RS1               NA                Complete               NA                    
4         1 repeat_separate                    1 NA                    mixed_structure_2 NA                RS1               NA                     Complete              
5         1 repeat_separate                    2 NA                    mixed_structure_1 RS2               NA                Complete               NA      
```

### Checklist
Before submitting this issue, please check and verify below that the submission meets the below criteria:

- [x] The issue is atomic
- [x] The issue description is documented
- [x] The issue title describes the problem succinctly
- [x] Developers are assigned to the issue
- [x] Labels are assigned to the issue

### LINKED ISSUE #206: [BUG] add_partial_keys() Doesn't fully capture arm regex
# Issue Description
The `add_partial_keys()` function is used to split events and arm names that REDCap provides as a single unique event name, ex: `repeat_event_arm_1`. `add_partial_keys()` will split this up into `repeat_event` and `1`.

An oversight we made is when REDCap events are very similarly named:

image

REDCap can also add additional specifiers that aren't just the numeric value. This means we need to capture that, and convert the `arm` column from an integer type to a factor type (so in the above example we capture `1b` as an `arm` value.

This will be corrected in the upcoming #205 

## How to Reproduce the Bug:
The issue occurs here, due to limited regex definition:
https://github.com/CHOP-CGTInformatics/REDCapTidieR/blob/82a49b713afff929c6d5df7230acab3085186b21/R/utils.R#L18-L34

### Checklist
Before submitting this issue, please check and verify below that the submission meets the below criteria:

- [x] The issue is atomic
- [x] The issue description is documented
- [x] The issue title describes the problem succinctly
- [x] Developers are assigned to the issue
- [x] Labels are assigned to the issue

### PR DESCRIPTION
**Pull request by rsh52**
# Description
This PR no longer implements `join_data_tibbles()`, that is being moved out into a new branch. Instead, this PR seeks to adress a few bugs, one minor (documented below), and one major where we are mistakenly reporting mixed data structure outputs.

I will be posting comments on the code itself below to explain some of the changes.

## The Missing Data Issue

Because we weren't making use of `redcap_event_instance` correctly, we were not accurately reporting all of the data associated with each REDCap event. See below from the Mixed Structure REDCap for reference:

image

And here is what currently comes out of the REDCapTidieR data tibbles from `main`:

<details>
  <summary><b>Current Output with Missing Data</b></summary>

```r
> sprtbl_mixed
# A REDCapTidieR Supertibble with 3 instruments
  redcap_form_name     redcap_form_label    redcap_data redcap_metadata   redcap_events structure data_rows data_cols
  <chr>                <chr>                <list>      <list>            <list>        <chr>         <int>     <int>
1 nonrepeat_form       Nonrepeat Form       <tibble>    <tibble [2 × 17]> <tibble>      nonrepea…         4         5
2 repeat_form          Repeat Form          <tibble>    <tibble [2 × 17]> <tibble>      mixed             2         5
3 mixed_structure_form Mixed Structure Form <tibble>    <tibble [2 × 17]> <tibble>      mixed             3         5
# ℹ 3 more variables: data_size <lbstr_by>, data_na_pct <formttbl>, form_complete_pct <formttbl>
> sprtbl_mixed$redcap_data
[[1]]
# A tibble: 4 × 5
  record_id redcap_event       redcap_event_instance nonrepeat_1 form_status_complete
      <dbl> <chr>                              <dbl> <chr>       <fct>               
1         1 non_repeating                         NA Nonrepeat 1 Incomplete          
2         1 repeating_separate                    NA Nonrepeat 2 Incomplete          
3         1 repeating_together                     1 A           Incomplete          
4         1 repeating_together                     2 B           Incomplete          

[[2]]
# A tibble: 2 × 5
  record_id redcap_event  redcap_form_instance repeat_1 form_status_complete
      <dbl> <chr>                        <dbl> <chr>    <fct>               
1         1 non_repeating                    1 Repeat 1 Incomplete          
2         1 non_repeating                    2 Repeat 2 Incomplete          

[[3]]
# A tibble: 3 × 5
  record_id redcap_event       redcap_form_instance mixed_structure_1 form_status_complete
      <dbl> <chr>                             <dbl> <chr>             <fct>               
1         1 non_repeating                         1 Mixed Nonrepeat 1 Incomplete          
2         1 repeating_separate                    1 Mixed Repeat 1    Incomplete          
3         1 repeating_separate                    2 Mixed Repeat 2    Incomplete     
```
</details>

In the second data tibble (the repeat form), we should expect to see 4 entries, 2 for the non repeating event due to 2 repeating form instances, and 2 for the repeating together event entries. 

In the third data tibble (the mixed structure form), we should expect to see 5 entries and are again missing the 2 repeating together event entries.

Here is the output with the proposed code changes:

<details>
  <summary><b>Proposed Output with Missing Data Added</b></summary>

```r
> sprtbl_mixed
# A REDCapTidieR Supertibble with 3 instruments
  redcap_form_name   redcap_form_label redcap_data redcap_metadata redcap_events structure data_rows data_cols data_size data_na_pct form_complete_pct
  <chr>              <chr>             <list>      <list>          <list>        <chr>         <int>     <int> <lbstr_b> <formttbl>  <formttbl>       
1 nonrepeat_form     Nonrepeat Form    <tibble>    <tibble>        <tibble>      nonrepea…         4         5 2.59 kB   0%          0%               
2 repeat_form        Repeat Form       <tibble>    <tibble>        <tibble>      mixed             4         6 2.67 kB   0%          0%               
3 mixed_structure_f… Mixed Structure … <tibble>    <tibble>        <tibble>      mixed             5         6 2.94 kB   0%          0%               
> sprtbl_mixed$redcap_data
[[1]]
# A tibble: 4 × 5
  record_id redcap_event       redcap_event_instance nonrepeat_1 form_status_complete
      <dbl> <chr>                              <dbl> <chr>       <fct>               
1         1 non_repeating                         NA Nonrepeat 1 Incomplete          
2         1 repeating_separate                    NA Nonrepeat 2 Incomplete          
3         1 repeating_together                     1 A           Incomplete          
4         1 repeating_together                     2 B           Incomplete          

[[2]]
# A tibble: 4 × 6
  record_id redcap_event       redcap_form_instance redcap_event_instance repeat_1 form_status_complete
      <dbl> <chr>                             <dbl>                 <dbl> <chr>    <fct>               
1         1 non_repeating                         1                    NA Repeat 1 Incomplete          
2         1 non_repeating                         2                    NA Repeat 2 Incomplete          
3         1 repeating_together                   NA                     1 A        Incomplete          
4         1 repeating_together                   NA                     2 B        Incomplete          

[[3]]
# A tibble: 5 × 6
  record_id redcap_event       redcap_form_instance redcap_event_instance mixed_structure_1 form_status_complete
      <dbl> <chr>                             <dbl>                 <dbl> <chr>             <fct>               
1         1 non_repeating                         1                    NA Mixed Nonrepeat 1 Incomplete          
2         1 repeating_separate                    1                    NA Mixed Repeat 1    Incomplete          
3         1 repeating_separate                    2                    NA Mixed Repeat 2    Incomplete          
4         1 repeating_together                   NA                     1 A                 Incomplete          
5         1 repeating_together                   NA                     2 B                 Incomplete        
```
</details>

# Proposed Changes 
List changes below in bullet format:

- Update `add_event_mapping()` with output from new function `get_repeat_event_types()` and make `repeat_type` available in the `redcap_events` column of the supertibble
- Update `clean_redcap_()`
   - Update `convert_mixed_instrument()` to detect RT data and shift instances over to `redcap_event_instance`
- Update `add_partial_keys()` to handle regex for arms that don't end in an integer (See #206)
- Update test suite accordingly
- ~Add new `join_data_tibbles()` function and tests~ <- in new branch

## Remaining TODO's

- Update NEWS.md
- Update pkgdown vignette
- ~Finish updating `join_data_tibbles()` to handle `by` appropriately for RT/RS data~
   - ~See comment in #204 about concerns~ <- <- in new branch

### Issue Addressed
**Closes #206**
**Closes #204**

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [NA] New/revised functions that update downstream outputs have associated static testing files (`.RDS`) updated under `inst/testdata/create_test_data.R`
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [ ] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR
- [NA] Pre-release package version incremented using `usethis::use_version()`
   - A part of the upcoming 1.2 release

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [ ] I checked that new files follow naming conventions and are in the right place
- [ ] I checked that documentation is complete, clear, and without typos
- [ ] I added/edited comments to explain "why" not "how"
- [ ] I checked that all new variable and function names follow naming conventions
- [ ] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [ ] I checked that new tests address important edge cases

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207922156958425

### CODE REVIEW

File: R/clean_redcap_long.R [hunk 1]
```diff
@@ -84,7 +84,9 @@ clean_redcap_long <- function(db_data_long,
       # Retrieve mixed structure fields and forms in reference df
       mixed_structure_ref <- get_mixed_structure_fields(db_data_long) %>%
         filter(.data$rep_and_nonrep & !str_ends(.data$field_name, "_form_complete")) %>%
-        left_join(db_metadata_long %>% select("field_name", "form_name"),
+        left_join(
```
Comment (rsh52) [inline R/clean_redcap_long.R:87]: No change, just lintr.

File: R/read_redcap.R [hunk 2]
```diff
@@ -292,7 +295,7 @@ read_redcap <- function(redcap_uri,
   out <- add_metadata(out, db_metadata, redcap_uri, token, suppress_redcapr_messages)
 
   if (is_longitudinal) {
-    out <- add_event_mapping(out, linked_arms)
+    out <- add_event_mapping(out, linked_arms, repeat_event_types)
```
Comment (rsh52) [inline R/read_redcap.R:298]: Add `repeat_event_types` reference table to `add_event_mapping()`

File: R/read_redcap.R [hunk 3]
```diff
@@ -440,18 +443,21 @@ add_metadata <- function(supertbl, db_metadata, redcap_uri, token, suppress_redc
...
 #'
 #' @return
 #' The original supertibble with an events \code{redcap_events} list column
 #' containing arms and events associated with each instrument
 #'
 #' @keywords internal
-#'
-add_event_mapping <- function(supertbl, linked_arms) {
+
+add_event_mapping <- function(supertbl, linked_arms, repeat_event_types) {
```
Comment (rsh52) [inline R/read_redcap.R:455]: `add_event_mapping()` now accepts a reference table for the repeating event types designations. This table is joined onto the `linked_arms` data to make `repeat_type` available in the eventual `redcap_events` supertibble column.

File: R/utils.R [hunk 4]
```diff
@@ -18,12 +18,13 @@
 add_partial_keys <- function(db_data,
                              var = NULL) {
   if (!is.null(enexpr(var))) {
-    pattern <- "^(\\w+?)_arm_(\\d)$"
+    # Include handling for instances where REDCap appends with "_1b" or similar
+    pattern <- "^(\\w+?)_arm_(\\d+\\w?)$"
```
Comment (rsh52) [inline R/utils.R:22]: Fix regex to address #206

File: R/utils.R [hunk 5]
```diff
@@ -18,12 +18,13 @@
...
 
     db_data <- db_data %>%
       mutate(
         redcap_event = sub(pattern, "\\1", {{ var }}),
-        redcap_arm   = as.integer(sub(pattern, "\\2", {{ var }}))
+        redcap_arm   = as.factor(sub(pattern, "\\2", {{ var }}))
```
Comment (rsh52) [inline R/utils.R:27]: Convert to factor to capture non-integer arm values as mentioned in #206

File: R/clean_redcap_long.R [hunk 6]
```diff
@@ -416,14 +421,38 @@ convert_mixed_instrument <- function(db_data_long, mixed_structure_ref) {
         )
       )
 
+    repeat_together_present <- any(
```
Comment (rsh52) [inline R/clean_redcap_long.R:424]: Since we now want to separate RT events so their instances get captured under `redcap_event_instance`, we first need to check for the right behavior for RTs (i.e. no repeating instrument but a repeat instance detected).

File: R/clean_redcap_long.R [hunk 7]
```diff
@@ -416,14 +421,38 @@ convert_mixed_instrument <- function(db_data_long, mixed_structure_ref) {
...
+
+    if (!"redcap_event_instance" %in% names(db_data_long) && repeat_together_present) {
+      db_data_long <- db_data_long %>%
+        mutate(
+          redcap_event_instance = NA
+        ) %>%
+        relocate(.data$redcap_event_instance, .after = .data$redcap_repeat_instance)
+    }
+
+    if (repeat_together_present) {
```
Comment (rsh52) [inline R/clean_redcap_long.R:437]: Finally, shift the redcap repeat instance values over to the redcap event instance column.

File: R/clean_redcap_long.R [hunk 8]
```diff
@@ -416,14 +421,38 @@ convert_mixed_instrument <- function(db_data_long, mixed_structure_ref) {
...
+          )
+        )
+    }
+
     # Assign update data based on rules below
     db_data_long <- db_data_long %>%
       mutate(
         redcap_repeat_instance = case_when(
-          # Add single instance repeat event instance vals when none exist
-          update_mask & is.na(redcap_repeat_instance) ~ 1,
```
Comment (rsh52) [inline R/clean_redcap_long.R:424]: I don't believe we need this any more, we will need to update the pkgdown though. Instead we need to have `redcap_repeat_instance` reflect the shift of the event instance values by setting them to `NA`.

File: tests/testthat/test-clean_redcap_long.R [hunk 9]
```diff
@@ -309,7 +309,7 @@ test_that("convert_mixed_instrument works", {
   expected_out <- tibble::tribble(
     ~record_id, ~redcap_repeat_instrument, ~redcap_repeat_instance, ~mixed_structure_variable,
     ~repeat_form_variable, ~mixed_repeat_var,
-    1, "mixed_structure_form", 1, "A", NA, NA,
+    1, "mixed_structure_form", NA, "A", NA, NA,
```
Comment (rsh52) [inline tests/testthat/test-clean_redcap_long.R:312]: See comment above about no longer using single instance for non repeat treatment.

Comment (rsh52) [reply tests/testthat/test-clean_redcap_long.R:312]: Not applicable, ignore.

File: R/clean_redcap_long.R [hunk 10]
```diff
@@ -318,7 +320,10 @@ distill_repeat_table_long <- function(form_name,
   db_data_long <- db_data_long %>%
     add_partial_keys(var = .data$redcap_event_name) %>%
     filter(
-      !is.na(.data$redcap_form_instance) &
+      (
+        !is.na(.data$redcap_form_instance) |
+          if_any(starts_with("redcap_event_instance"), ~ !is.na(.))
+      ) &
```
Comment (rsh52) [inline R/clean_redcap_long.R:326]: Filter for when either `redcap_form_instance` is `NA` or `redcap_event_instance` is `NA` (if it exists).

File: R/join_data_tibbles.R [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+#' @inheritParams join_data_tibbles
+#' @param join_fn the user specified join function type output by [get_join_fn()]
+#' @param is_mixed TRUE/FALSE mixed data structure
+#'
+#' @returns a dataframe
+#'
+#' @keywords internal
+
+join_tbls <- function(x, y, join_fn, by, suffix, is_mixed) {
+  if (is_mixed) {
```
Comment (rsh52) [inline R/join_data_tibbles.R:187]: TBD if this is will be used pending conversation in #204

File: R/utils.R (hunk 5 above)
Comment (ezraporter) [reply R/utils.R:27]: I think `character` is probably better here since these will usually be numbers and aren't the type of categorical variables you normally expect as factor to represent

**Review by ezraporter — CHANGES_REQUESTED** [review]
One tiny comment but looks good otherwise

**Comment by rsh52** [comment]
> I think `character` is probably better here since these will usually be numbers and aren't the type of categorical variables you normally expect as factor to represent

Good call, thanks. I thought a factor would be better for how `redcap_arm` gets handled later but it doesn't seem to really impact anything and you're right this is likely more representative.

**Review by ezraporter — APPROVED** [review]
